Scripting: Convert script fields to use script context

rjernst opened the pull request:
This commit removes the use of SearchScript for script fields and adds
a new FieldScript.

jdconrad commented on docs/painless/painless-contexts/painless-field-context.asciidoc:
```diff
@@ -14,18 +14,15 @@ a customized value for each document in the results of a query.
         `List` of values.
 
-{ref}/mapping-source-field.html[`ctx['_source']`] (`Map`)::
+{ref}/mapping-source-field.html[`params['_source']`] (`Map`)::
```
Do you mind adding read-only to this, like how _score was?  This shouldn't be modified as part of a script field.  My mistake from the first iteration of these docs.

jdconrad commented on server/src/main/java/org/elasticsearch/script/FieldScript.java:
```diff
@@ -0,0 +1,100 @@
+        Map<String, String> deprecations = new HashMap<>();
+        deprecations.put(
+            "doc",
+            "Accessing variable [doc] via [params.doc] from within a terms-set-query-script " +
```
Think this was a copy and paste issue -- should be field-script and not terms-set-query

rjernst commented:
@jdconrad I addressed your comments. I will wait for https://github.com/elastic/elasticsearch/pull/33856 to get in so I can re-use the expression script engine work there to add expression support of the new field script.
